File: rc-application/src/main/java/com/transkript/reportcard/business/service/i/SubjectRegistrationService.java
package com.transkript.reportcard.business.service.i;

import com.transkript.reportcard.api.dto.SubjectRegistrationDto;
import com.transkript.reportcard.api.dto.response.EntityResponse;
import com.transkript.reportcard.data.entity.SubjectRegistration;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.springframework.transaction.annotation.Transactional;

public interface SubjectRegistrationService {
    @NotNull
    EntityResponse create(@NotNull SubjectRegistrationDto subjectRegistrationDto);

    @NotNull
    List<EntityResponse> createMultiple(@NotNull List<SubjectRegistrationDto> subjectRegistrationDtoList);

    @NotNull
    List<SubjectRegistrationDto> getAllDtoBySAT(Long satId);

    @NotNull
    SubjectRegistrationDto getDto(Long id);

    @Transactional
    void delete(Long id);

    @NotNull
    SubjectRegistration getEntity(Long id);

    @NotNull
    SubjectRegistration getEntityBySubjectAndSat(Long subjectId, Long satId);
}

File: rc-application/src/main/java/com/transkript/reportcard/business/service/i/SubjectRegistrationService.kt
package com.transkript.reportcard.business.service.i

import com.transkript.reportcard.api.dto.SubjectRegistrationDto
import com.transkript.reportcard.api.dto.response.EntityResponse
import com.transkript.reportcard.data.entity.SubjectRegistration
import org.springframework.transaction.annotation.Transactional

interface SubjectRegistrationService {
    fun create(subjectRegistrationDto: SubjectRegistrationDto): EntityResponse
    fun createMultiple(subjectRegistrationDtoList: List<SubjectRegistrationDto>): List<EntityResponse>
    fun getDtoList(satId: Long): List<SubjectRegistrationDto>
    fun getDto(registrationId: Long): SubjectRegistrationDto

    @Transactional
    fun delete(registrationId: Long)
    fun getEntity(registrationId: Long): SubjectRegistration
    fun getEntity(satId: Long, subjectId: Long): SubjectRegistration
    fun getEntitiesByApplicationTrial(satId: Long): List<SubjectRegistration>
}
